### PR TITLE
[animations] adds alignment parameter to FadeScaleTransition

### PR DESCRIPTION
**Pull request by grinder15**
Implements Fade scale anchor point variant as documented here: https://material.io/design/motion/the-motion-system.html#fade

An example which the menu scales from top right when fading in.
menu

### CODE REVIEW

File: packages/animations/lib/src/fade_scale_transition.dart [hunk 1]
```diff
@@ -123,6 +123,7 @@ class FadeScaleTransition extends StatelessWidget {
   const FadeScaleTransition({
     Key key,
     @required this.animation,
+    this.alignment = Alignment.center,
```
Comment (goderbauer) [inline packages/animations/lib/src/fade_scale_transition.dart:126]: Update documentation above that alignment cannot be null and add an assert below.

File: packages/animations/test/fade_scale_transition_test.dart [hunk 2]
```diff
@@ -146,27 +156,35 @@ void main() {
...
+      expect(topScaleAlignment, Alignment.center);
 
       // End of opacity animation
       await tester.pump(const Duration(milliseconds: 25));
       topFadeTransitionOpacity = _getOpacity(key, tester);
       expect(topFadeTransitionOpacity, 0.0);
       topScale = _getScale(key, tester);
       expect(topScale, 1.0);
+      topScaleAlignment = _getScaleAlignment(key, tester);
+      expect(topScaleAlignment, Alignment.center);
```
Comment (goderbauer) [inline packages/animations/test/fade_scale_transition_test.dart:187]: All tests seem to be using `Alignment.center`. There should also be one that doesn't use the default.

Comment (grinder15) [reply packages/animations/test/fade_scale_transition_test.dart:187]: so in order to modify `FadeScaleTransitionConfiguration`, I need to extend it and override `transitionBuilder` with different `FadeScaleTransition.alignment`. Is it okay to put the extended configuration in that test?

Comment (shihaohong) [reply packages/animations/test/fade_scale_transition_test.dart:187]: I think that's, unfortunately, the only way we can have FadeScaleTransition properly make use of the alignment parameter. I'm just worried about its ease-of-use since the user would have to extend the `FadeScaleTransitionConfiguration` widget, as you mentioned. 

I still think it's okay to do it, but we'll probably have to add something in the docs so that users are aware how to achieve this effect.

File: packages/animations/lib/src/fade_scale_transition.dart [hunk 3]
```diff
@@ -135,6 +139,19 @@ class FadeScaleTransition extends StatelessWidget {
...
+  ///
+  /// By default, the child scales from it's center when fading in. In a
+  /// variation on the default behavior, the scale anchor point can be
+  /// repositioned from the center of a child to another location within
+  /// the child.
+  ///
+  /// For example, when an overflow icon (the parent component) triggers a menu
+  /// (the child component), the menu's anchor point can be positioned at the
+  /// top right corner of the menu so that it appears to scale from the overflow
+  /// icon during a transition.
```
Comment (shihaohong) [inline packages/animations/lib/src/fade_scale_transition.dart:152]:
```dart
/// To modify alignment, `FadeScaleTransitionConfiguration` will 
/// need to be extended and this new configuration class will need to be used instead:
/// ```
/// class _ModifiedFadeScaleTransitionConfiguration extends FadeScaleTransitionConfiguration {
///   const _ModifiedFadeScaleTransitionConfiguration();
///
///   @override
///   Widget transitionBuilder(
///     BuildContext context,
///     Animation<double> animation,
///     Animation<double> secondaryAnimation, 
///     Widget child,
///   ) {
///     return FadeScaleTransition(
///       animation: animation,
///       alignment: Alignment.topRight, // add this line
///       child: child,
///     );
///   }
/// }
/// ```
```

File: packages/animations/test/fade_scale_transition_test.dart [hunk 4]
```diff
@@ -71,8 +71,10 @@ void main() {
       // Opacity duration: First 30% of 150ms, linear transition
       double topFadeTransitionOpacity = _getOpacity(key, tester);
       double topScale = _getScale(key, tester);
+      Alignment topScaleAlignment = _getScaleAlignment(key, tester);
```
Comment (shihaohong) [inline packages/animations/test/fade_scale_transition_test.dart:74]: Since we don't anticipate the `topScaleAlignment` to change throughout the animation (it doesn't really make sense for it to), maybe it's better to revert these changes to the existing tests and add one, simpler test that verifies that the `ScaleTransition`'s alignment was set to the appropriate value when the modal is shown.

File: packages/animations/test/fade_scale_transition_test.dart (hunk 2 above)
Comment (grinder15) [reply packages/animations/test/fade_scale_transition_test.dart:187]: I was thinking if I would add `alignment`/`scaleAlignment` param for `FadeScaleTransitionConfiguration`  to avoid extending it but since the motion system docs only shows dialog/modal scales at center, I think it doesn't need that param, even some docs on achieving that effect on dialogs/modals.

I just asking because I want to follow the same test schemes on `fade_scale_test.dart`, using a modal to test `FadeScaleTransition`.

**Comment by grinder15** [comment]
@shihaohong 
Just to be clear, the goal of this change was to add Control to scale animation component of `FadeScaleTransition` widget, not the configuration of a modal using this transition widget. 😃  ohh and the example I posted above was a butchered version of `PopupMenuButton` using `FadeScaleTransition` with scale alignment at topRight to present the use case of this PR, not a modal positioned at topRight with transparent scrim color. 

Well, I could also release my versions of `SnackBar` and `PopupMenuButton` implemented using this widget after this PR, its just my code is not yet clean and those private classes to implement those widgets give me headaches.

File: packages/animations/lib/src/fade_scale_transition.dart (hunk 3 above)
Comment (grinder15) [reply packages/animations/lib/src/fade_scale_transition.dart:152]: I'll add this doc to `FadeScaleTransitionConfiguration` class instead.

**Comment by shihaohong** [comment]
Sorry for taking so long to get back to this.

I think we should hold off on exposing the alignment API in this way. It doesn't seem very user-friendly for two reasons: 

1) It required an extension of the transition, which will then need to be applied to whichever widget it needs to transition, and
2) As you mentioned, it works with your unique version of PopupMenuButton, but does not expose a general, clean way to achieve the fade scale anchor point transition to other widgets. 

I would be more comfortable with accepting the PR if there was a design proposal for how to get this transition variant working for a general use case for the fade scale anchor point variant, maybe in the form of a GitHub issue or a design document.

I'm going to close the PR for now, but feel free to respond and re-open it if you have an idea for how to make this feature more usable!
